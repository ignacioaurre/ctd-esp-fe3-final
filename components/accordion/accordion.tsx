import React, {FC} from 'react'
import Accordion from '@mui/material/Accordion'
import AccordionDetails from '@mui/material/AccordionDetails'
import AccordionSummary from '@mui/material/AccordionSummary'
import Typography from '@mui/material/Typography'
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import Link from 'next/link'

type Items = {
  resourceURI: string,
  name: string,
}

type Characters = {
  available: number,
  collectionURI: string,
  items: Items[] ,
  returned: number,
}

type AccordionProps = {
    id: number,
    title: string,
    description?: string,
    characters?: Characters,
}

const AccordionComponent: FC<AccordionProps> = ({id, title, description , characters}: AccordionProps) => {
  
  const descriptionMsg = description === '' || null ? "No hay descripciÃ³n" : description;
  const charactersMsg = "No hay personajes para este comic";

  const urlNormalizer = (url: String) => {
    const newUrl = url?.split("/").slice(-1)[0];
    const baseUrl = url?.split("/").slice(-2)[0] == "characters" ? "personajes" : "comic"
    return `/${baseUrl}/${newUrl}`
  }
  
  return (
    <Accordion key={id}>
        <AccordionSummary
        expandIcon={<ExpandMoreIcon />}
        aria-controls="panel1a-content"
        id="panel1a-header"
        >
        <Typography>{title}</Typography>
        </AccordionSummary>
        <AccordionDetails sx={{backgroundColor: "whitesmoke"}}>
        {!characters ?  <Typography>{descriptionMsg}</Typography>
        : characters.available > 0 ? (
          characters.items.map(char => <Link href={urlNormalizer(char.resourceURI)} key={char.name} style={{display: "block"}}><p style={{cursor: 'pointer'}}>{char.name}</p></Link>)
        ): <Typography>{charactersMsg}</Typography>}
        </AccordionDetails>
    </Accordion>
  )
}

export default AccordionComponent;