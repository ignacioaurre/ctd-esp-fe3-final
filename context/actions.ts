import { Card, Delivery, Register, ComicInfo } from 'dh-marvel/features/Types/state.types';
import { Dispatch } from 'react';
import { Action } from './reducer';

export const selectComic = (dispatch: Dispatch<Action>, comic: ComicInfo) => {
    dispatch({type: 'ADD_COMIC', payload: comic})
}

export const submitRegister = (dispatch: Dispatch<Action>, register: Register) => {
    dispatch({type: 'SUBMIT_REGISTER', payload: register});
}

export const submitDelivery = (dispatch: Dispatch<Action>, delivery: Delivery) => {
    dispatch({type: 'SUBMIT_DELIVERY', payload: delivery});
}

export const submitCard = (dispatch: Dispatch<Action>, card: Card) => {
    dispatch({type: 'SUBMIT_CARD', payload: card});
}

export const submitForm = (dispatch: Dispatch<Action>) => {
    dispatch({ type: 'SUBMIT_FORM'})
};

export const restoreState = (dispatch: Dispatch<Action>) => {
    dispatch({ type: 'RESTORE_STATE' })
}

export const setSnackbar = (dispatch: Dispatch<Action>, snackbar: string) => {
    dispatch({ type: 'SET_SNACKBAR', payload: snackbar})
}