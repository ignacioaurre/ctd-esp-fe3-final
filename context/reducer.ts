import { OrderState } from "context";
import { Comic } from "dh-marvel/features/Types/comic.types"
import { Card, Delivery, Register, Order, ComicInfo } from "dh-marvel/features/Types/state.types"

export const types = {
    ADD_COMIC: 'ADD_COMIC',
    SUBMIT_REGISTER: 'SUBMIT_REGISTER',
    SUBMIT_DELIVERY: 'SUBMIT_DELIVERY',
    SUBMIT_CARD: 'SUBMIT_CARD', 
}

export type Action = 
  |  { type: 'ADD_COMIC' , payload: ComicInfo}
  |  { type: 'SUBMIT_REGISTER', payload: Register }
  |  { type: 'SUBMIT_DELIVERY', payload: Delivery }
  |  { type: 'SUBMIT_CARD', payload: Card };

export const initialState: OrderState = { 
    order: {
        comic: {
            img: "",
            price: 0,
            title: "",
        },
        register: {
            usuario: "",
            apellido: "",
            email: "",
        },
        delivery: {
            direccion: "Siempre Viva 123",
            departamento: "",
            ciudad: "",
            provincia: "",
            codigoPostal: "",
        },
        card: {
            nroTarjeta: "",
            nombreTarjeta: "",
            fechaExp: "",
            codSeguridad: "",
        }
    }
}

export const reducer = (state: OrderState, action: Action) => {
    switch (action.type) {
        case 'ADD_COMIC':
            return {...state, comic: action.payload}
        case 'SUBMIT_REGISTER':
            return {...state, register: action.payload}
        case 'SUBMIT_DELIVERY':
            return {...state, delivery: action.payload}
        case 'SUBMIT_CARD':
            return {...state, card: action.payload}
        default:
      return state;
    }
}