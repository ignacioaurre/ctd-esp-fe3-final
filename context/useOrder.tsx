import { OrderContext, OrderContextState } from "context";
import { useContext } from "react";

const useOrder = (): OrderContextState => {
    const context = useContext(OrderContext);
    if (!context) {
        throw new Error(
            'useOrder must be used within a OrderProvider'
        );
    }
    return context;
};

export default useOrder;