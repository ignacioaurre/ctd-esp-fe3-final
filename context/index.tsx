import { Order } from 'dh-marvel/features/Types/state.types';
import React, { useMemo, useReducer, createContext, FC, Dispatch, PropsWithChildren } from 'react';

import { reducer, initialState, Action } from './reducer';

export interface OrderState {
  order: Order
}

export type OrderContextState = {
  state: {order: Order},
  dispatch: Dispatch<Action>
}

export const OrderContext = createContext<OrderContextState | undefined>(undefined);

const ContextProvider: FC<PropsWithChildren> = ({ children }: PropsWithChildren ) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const contextValue = useMemo(
    () => ({
      state,
      dispatch,
    }),
    [state, dispatch],
  );

  return (
        <OrderContext.Provider value={contextValue} >
          {children}
        </OrderContext.Provider>
      )

};

export default ContextProvider;