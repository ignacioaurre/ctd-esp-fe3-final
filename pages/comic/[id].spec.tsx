import {render, screen, waitFor} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import useOrder from 'dh-marvel/context/useOrder';
import ContextProvider from 'dh-marvel/context/index';
import { initialState } from 'dh-marvel/context/reducer';
import ComicDetail from './[id].page';
import { Characters } from 'dh-marvel/components/accordion/accordion';
import { useRouter } from 'next/router';

jest.mock("dh-marvel/context/useOrder")
const mockUseOrder = useOrder as jest.MockedFunction<typeof useOrder>
const mockDispatch = jest.fn();
mockUseOrder.mockReturnValue({
    state: initialState, 
    dispatch: mockDispatch
})
const mockPush = jest.fn();
jest.mock('next/router', () => ({
    useRouter: jest.fn(),
}));
(useRouter as jest.Mock).mockImplementation(() => ({
    push: mockPush,
}))

const props = {
    id: 1,
    thumbnail: {
        path: "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
        extension: "jpg",
    },
    pageCount: 4,
    title: "titulo",
    description: "descripbion",
    characters: {} as Characters,
    price: 24,
    oldPrice: 20,
    stock: 2,
    format: "format",
}

describe("ID comic Page", () => {
    test("Component should match snapshot", () => {
        const { baseElement } = render (
            <ContextProvider>
            <ComicDetail {...props} />
        </ContextProvider>
        )
        expect(baseElement).toMatchSnapshot();
    })
    test("Button should be clickeable for purchase", async () => {
        render (
            <ContextProvider>
            <ComicDetail {...props} />
        </ContextProvider>
        )

        const purchaseButton = screen.getByRole('button', {name: /Comprar/i,});
        expect(purchaseButton).not.toBeDisabled();
        await userEvent.click(purchaseButton);

        expect(mockDispatch).toBeCalledWith({
            payload: {
                img: "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg",
                price: 24,
                title: "titulo",
            },
            type: "ADD_COMIC"
        })
    })
})
