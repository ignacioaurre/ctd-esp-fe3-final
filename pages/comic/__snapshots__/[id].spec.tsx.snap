// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ID comic Page Component should match snapshot 1`] = `
<body>
  <div>
    <div
      class="css-1rzuxdl-MuiStack-root"
    >
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular css-r6ewbb-MuiToolbar-root"
          >
            <a
              class="MuiTypography-root MuiTypography-h6 MuiLink-root MuiLink-underlineAlways css-973mah-MuiTypography-root-MuiLink-root"
              href="/"
            >
               DH-Marvel
            </a>
            <div
              class="MuiBox-root css-0"
            >
              <a
                class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineAlways css-f11se6-MuiTypography-root-MuiLink-root"
                href="/faqs"
              >
                FAQ
              </a>
            </div>
          </div>
        </div>
      </header>
      <div
        class="MuiBox-root css-1hog1e3"
      >
        <div
          class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
        >
          <div
            class="css-fdaco8-MuiStack-root"
          >
            <h2
              class="MuiTypography-root MuiTypography-h2 css-14leyog-MuiTypography-root"
            >
              Detalle del Comic
            </h2>
            <div
              class="MuiGrid2-root MuiGrid2-container MuiGrid2-spacing-xs-2 css-1nwx68v-MuiGrid2-root"
            >
              <div
                class="MuiGrid2-root MuiGrid2-grid-xs-3 css-p4z6b7-MuiGrid2-root"
              >
                <span
                  style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                >
                  <span
                    style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27200%27%20height=%27300%27/%3e"
                      style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                    />
                  </span>
                  <img
                    alt="Portada del comic"
                    data-nimg="intrinsic"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </span>
              </div>
              <div
                class="MuiGrid2-root MuiGrid2-grid-xs-3 css-p4z6b7-MuiGrid2-root"
              >
                <div
                  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-830i6s-MuiPaper-root-MuiCard-root"
                >
                  <div
                    class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-16m3jxf-MuiTypography-root"
                    >
                      format
                    </p>
                    <div
                      class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
                    >
                      titulo
                    </div>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-170qu9w-MuiTypography-root"
                    >
                      $20
                    </p>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      $
                      24
                    </p>
                  </div>
                  <div
                    class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
                  >
                    <button
                      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge css-189e9pj-MuiButtonBase-root-MuiButton-root"
                      tabindex="0"
                      type="button"
                    >
                      Comprar
                       
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid2-root MuiGrid2-grid-xs-7 css-1knq2yb-MuiGrid2-root"
              >
                <div
                  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1elwnq4-MuiPaper-root-MuiAccordion-root"
                >
                  <div
                    aria-controls="panel1a-content"
                    aria-expanded="false"
                    class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
                    id="panel1a-header"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                      >
                        Descripcion
                      </p>
                    </div>
                    <div
                      class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="ExpandMoreIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
                    style="min-height: 0px;"
                  >
                    <div
                      class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
                    >
                      <div
                        class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
                      >
                        <div
                          aria-labelledby="panel1a-header"
                          class="MuiAccordion-region"
                          id="panel1a-content"
                          role="region"
                        >
                          <div
                            class="MuiAccordionDetails-root css-zpvdp-MuiAccordionDetails-root"
                          >
                            <p
                              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                            >
                              descripbion
                            </p>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid2-root MuiGrid2-grid-xs-7 css-1knq2yb-MuiGrid2-root"
              >
                <div
                  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1elwnq4-MuiPaper-root-MuiAccordion-root"
                >
                  <div
                    aria-controls="panel1a-content"
                    aria-expanded="false"
                    class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
                    id="panel1a-header"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                      >
                        Personajes
                      </p>
                    </div>
                    <div
                      class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="ExpandMoreIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
                    style="min-height: 0px;"
                  >
                    <div
                      class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
                    >
                      <div
                        class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
                      >
                        <div
                          aria-labelledby="panel1a-header"
                          class="MuiAccordion-region"
                          id="panel1a-content"
                          role="region"
                        >
                          <div
                            class="MuiAccordionDetails-root css-zpvdp-MuiAccordionDetails-root"
                          >
                            <p
                              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                            >
                              No hay personajes para este comic
                            </p>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <footer
        class="MuiBox-root css-2pi20k"
      >
        <a
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-zfdr7-MuiTypography-root-MuiLink-root"
          href="https://www.digitalhouse.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          Powered by
           
          <div
            class="MuiBox-root css-18lsnf1"
          >
            <span
              style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
            >
              <span
                style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27168%27%20height=%2713%27/%3e"
                  style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                />
              </span>
              <img
                alt="Digital House Logo"
                data-nimg="intrinsic"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
          </div>
        </a>
      </footer>
    </div>
  </div>
</body>
`;
